clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -187,15 +189,22 @@ void setTickRate(Duration tickRate) {
 
     private void timeoutHandshakeResponse(long timeout, TimeUnit unit) {
         ScheduledExecutorService scheduledThreadPool = Executors.newSingleThreadScheduledExecutor();
-        scheduledThreadPool.schedule(() -> handshakeResponseFuture.completeExceptionally(
-                new TimeoutException("Timed out waiting for the server to respond to the handshake message.")), timeout, unit);
+        scheduledThreadPool.schedule(() -> {
+            // If onError is called on a completed subject the global error handler is called
+            // REVIEW: Should we lock?
```

Comment:
So the worst case scenario here is:

1. Handshake Response arrives *barely* in time
1. Subject is completed AFTER this code checks `hasComplete`
1. Immediately after being completed, the Subject is `onError`ed by the timeout
1. Start completes because the subject was completed
1. Global onError handler is fired with the timeout error.

Technically there's nothing that will "fail" the process. The handler is called spuriously but the actual action succeeds. If the race goes the other way (timeout trips before the subject is completed) then so be it (that race exists in .NET and we're fine with it there)

I'm OK with that rare race, at least until we get feedback about it happening in the wild. The locking would be a bit rough here since we'd need to lock around all completion actions.

Reply:
Yeah, I was leaning towards not worrying about it.